c++/kernel_overhead_cl/oclutils.cpp

```diff
@@ -0,0 +1,270 @@
+{
+    cl::Program::Sources sources;
+    for (vector<string>::const_iterator it = names.begin(); it != names.end(); ++it) {
+        ifstream in((boost::format("%s/%s") % OpenCLUtils::getKernelDir() % *it).str().c_str());
+        if (!in.good())
+            throw runtime_error((boost::format("failed to open kernel file >%s<") % *it).str());
+        string result((istreambuf_iterator<char>(in)), istreambuf_iterator<char>());
+        char *buf = (char *)malloc(result.size() + 1);
+        strcpy(buf, result.c_str());
+        sources.push_back(std::make_pair(buf, result.size()));
+    }
+
+    return sources;
+}
+
+/**
+ * Prints the build log of a program on stderr.
+ * @param program Input: The program object
+ * @param devices Input: The device objects
+ */
+static void printProgramBuildLog(const cl::Program *program, const vector<cl::Device> &devices)
+{
+    string log;
+    log.resize(2048);
+    for (int i = 0; i < devices.size(); ++i) {
+        cerr << "============ build log for device " << i << ": ============\n";
+        cl_int error;
+        const string log = program->getBuildInfo<CL_PROGRAM_BUILD_LOG>(devices[i], &error);
+        CL_CHECK(error);
+        cerr << log << endl;
+    }
+}
+
+void OpenCLUtils::init(const vector<pair<string, string> > &sources, cl_device_type deviceType, const string &buildOptions)
+{
+    cl_int error;
+
+    // --- BEGIN get device and create context ---------------------------
+    // get platforms
+    platforms.clear();
+    getPlatforms(&platforms);
+    if (platforms.empty())
+        throw runtime_error("No OpenCL platform found");
+
+    // get device
+    devices.clear();
+    for (pfmIndex = 0; pfmIndex < platforms.size(); ++pfmIndex) {
+        platforms[pfmIndex].getDevices(deviceType, &devices);
+        if (!devices.empty())
+            break; // found at last one relevant device on this platform
+    }
+
+    // try default device, if specified deviceType is not found
+    if (pfmIndex == platforms.size()) {
+        cerr << "Could not find a device of type " << deviceType << "; using default device!" << endl;
+        for (pfmIndex = 0; pfmIndex < platforms.size(); ++pfmIndex) {
+            platforms[pfmIndex].getDevices(CL_DEVICE_TYPE_DEFAULT, &devices);
+            if (!devices.empty())
+                break; // found at last one relevant device on this platform
+            }
+    }
+
+    if (pfmIndex == platforms.size())
+        throw runtime_error("No relevant OpenCL devices found on any platform");
+
+    devIndex = 0; // for now, use first device
+    const string deviceName = getDeviceName(devices[devIndex]);
+    const string platformName = getPlatformName(platforms[pfmIndex]);
+
+    // create context
+    cl_context_properties contextProperties[] =
+    {
+        CL_CONTEXT_PLATFORM, (cl_context_properties)(platforms[pfmIndex])(),
+        0
+    };
+    context.reset(new cl::Context(devices, contextProperties, errorCallback, 0, &error));
```

Comment:
Here is where you should add the proper flag to enable profiling info
CL_QUEUE_PROFILING_ENABLE